Enable rootless AppArmor

kernelmethod opened the pull request:
<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

Related issues:
- https://github.com/containers/common/issues/958 (closed in https://github.com/containers/common/pull/960, reverted by https://github.com/containers/common/pull/969)
- https://github.com/containers/podman/issues/15874

As previously noted in https://github.com/containers/common/issues/958 and https://github.com/containers/podman/issues/15874, it is perfectly possible for rootless containers to run with an AppArmor profile. However, the current architecture of Podman requires it to load the default AppArmor profile at runtime before running the container.

With this PR, we no longer attempt to load the profile at runtime. Instead (as discussed in https://github.com/containers/podman/issues/15874), default profiles should be installed to `/etc/apparmor.d/` at the point where Podman is installed, and then specified in a default `containers.conf`.

This is the sister PR to https://github.com/containers/common/pull/1572; they should be merged in together to avoid breakage.

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

kernelmethod committed:
Allow rootless containers to use AppArmor profiles

These changes make it possible to specify an AppArmor profile to run a
container under via the --security-opt apparmor=... flag when rootless.
Podman previously required root privileges to specify an AppArmor in
order to check /sys/kernel/security/apparmor/profiles to see whether or
not the provided AppArmor profile was loaded into the kernel. After
these changes, Podman will instead error out when it attempts to write
/proc/thread-self/attr/exec with a non-existent profile.

Note that Podman can't use apparmor_parser to load a new profile into
the kernel without root. There isn't a convenient way to allow
generateSpec to run CheckProfileAndLoadDefault rootlessly from
containers/common in its current implementation, so I've removed all
uses of that function from generateSpec (even when we're running as
root). In theory this would mean that under these changes we won't be
able to load a default profile for containers. However, it turns out
that in practice there was never any path through the code where Podman
would load the default profile anyways, so this is a non-breaking
change.

Fixes an issue previously described in:
https://github.com/containers/common/issues/958

Signed-off-by: Will Shand <wss2ec@g.ucla.edu>

kernelmethod committed:
Stop expecting rootless AppArmor test to fail in test_podman_baseline.sh

Signed-off-by: Will Shand <wss2ec@g.ucla.edu>

rhatdan commented on test/e2e/run_apparmor_test.go:
```diff
@@ -43,6 +50,8 @@ var _ = Describe("Podman run", func() {
 
 	It("podman run no apparmor --privileged", func() {
 		skipIfAppArmorDisabled()
+		// Root is required to use --privileged
```
Why?

kernelmethod replied:
Oops, this one was a mistake. Fixed.

rhatdan commented on libpod/container_internal_common.go:
```diff
@@ -50,6 +49,7 @@ import (
 	"github.com/containers/storage/pkg/lockfile"
 	stypes "github.com/containers/storage/types"
 	securejoin "github.com/cyphar/filepath-securejoin"
+	runcaa "github.com/opencontainers/runc/libcontainer/apparmor"
```
I would rather keep the name apparmor, I think that is easier for future developers to understand.

kernelmethod committed:
Small fixes to https://github.com/containers/podman/pull/19303

- Import github.com/opencontainers/runc/libcontainer/apparmor as
  apparmor rather than runcaa for clarity.
- Don't skip the 'podman run no apparmor --privileged' if we're running
  rootlessly.

Signed-off-by: Will Shand <wss2ec@g.ucla.edu>

rhatdan commented:
@kernelmethod Still working on this one?

kernelmethod commented:
I am, I've just fallen behind on getting it finished. If you'd prefer that I close this PR and open a new one when it's ready to merge, that's fine with me.

rhatdan commented:
No problem, I would prefer you update this PR with fixes when you get a chance.

rhatdan commented:
@kernelmethod any update?

kernelmethod commented:
Not at the moment; I'm planning on returning to this in mid- / late-October. If somebody else wants to pick this up then they're free to do so, but it's currently low-priority for me.
